cmd/disk-cache.go

```diff
@@ -0,0 +1,764 @@
+	GetBucketInfoFn           func(bucket string) (bucketInfo BucketInfo, err error)
+	NewMultipartUploadFn      func(bucket, object string, metadata map[string]string) (uploadID string, err error)
+	PutObjectPartFn           func(bucket, object, uploadID string, partID int, data *hash.Reader) (info PartInfo, err error)
+	AbortMultipartUploadFn    func(bucket, object, uploadID string) error
+	CompleteMultipartUploadFn func(bucket, object, uploadID string, uploadedParts []CompletePart) (objInfo ObjectInfo, err error)
+}
+
+// CacheObjectLayer implements primitives for cache object API layer.
+type CacheObjectLayer interface {
+	// Bucket operations.
+	ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error)
+	GetBucketInfo(bucket string) (bucketInfo BucketInfo, err error)
+	ListBuckets() (buckets []BucketInfo, err error)
+
+	// Object operations.
+	GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer, etag string) (err error)
+	GetObjectInfo(bucket, object string) (objInfo ObjectInfo, err error)
+	PutObject(bucket, object string, data *hash.Reader, metadata map[string]string) (objInfo ObjectInfo, err error)
+
+	// Multipart operations.
+	NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error)
+	PutObjectPart(bucket, object, uploadID string, partID int, data *hash.Reader) (info PartInfo, err error)
+	AbortMultipartUpload(bucket, object, uploadID string) error
+	CompleteMultipartUpload(bucket, object, uploadID string, uploadedParts []CompletePart) (objInfo ObjectInfo, err error)
+}
+
+// choose cache deterministically based on hash of bucket,object. The hash index is treated as
+// a hint. In the event that the cache drive at hash index is offline, treat the list of cache drives
+// as a circular buffer and walk through them starting at hash index until an online drive is found.
+// if locate flag is set to true, do a stat to find drive where object is actually cached. If
+// object is not found, fall back to the first online cache drive closest to the hash index.
+func (c diskCache) getCacheFS(bucket, object string, locate bool) (*cacheFSObjects, error) {
+
+	index := c.hashIndex(bucket, object)
+	numDisks := len(c.cfs)
+	i := index
+	// save first online cache disk closest to the hint index
+	var matchDisk *cacheFSObjects
+	firstMatchDisk := false
+	for {
+		if c.cfs[i].getOnline() {
+			// if locate is set to true, save the first available disk closest to the index
+			if locate && !firstMatchDisk {
+				firstMatchDisk = true
+				matchDisk = c.cfs[i]
+			}
+			// if locate is set to true, check if object is cached to the disk at index i
+			if !locate || (locate && c.cfs[i].Exists(bucket, object)) {
+				return c.cfs[i], nil
+			}
+		}
+		i = (i + 1) % numDisks
+		if i == index {
+			break
+		}
+	}
+	if locate && matchDisk != nil {
+		return matchDisk, nil
+	}
+	return nil, errDiskNotFound
+}
```

Comment:
If you break the above function into two functions, its more readable and easy on the eyes:
```
func (c diskCache) getCacheFSRead(bucket, object string) (*cacheFSObjects, error) {
        index := c.hashIndex(bucket, object)
        numDisks := len(c.cfs)
        i := index
        var firstOnlineDisk *cacheObjects
        for {
                if c.cfs[i].IsOnline() {
                        if firstOnlineDisk == nil {
                                firstOnlineDisk = c.cfs[i]
                        }
                        if c.cfs[i].Exists(bukcet, object) {
                                return c.cfs[i], nil
                        }
                }
                i = (i + 1) % numDisks
                if i == index {
                        break
                }
        }
        if firstOnlineDisk != nil {
                return firstOnlineDisk, nil
        }
        return nil, errDiskNotFound
}

func (c diskCache) getCacheFSWrite(bucket, object string) (*cacheFSObjects, error) {
        index := c.hashIndex(bucket, object)
        numDisks := len(c.cfs)
        i := index

        for {
                if c.cfs[i].IsOnline() {
                        return c.cfs[i], nil
                }
                i = (i + 1) % numDisks
                if i == index {
                        break
                }
        }
        return nil, errDiskNotFound
}
```